File: dsa/queues/ruby/linked_list_queue.rb
require 'test/unit'
require 'byebug'

class Node
  attr_accessor :value, :next, :prev

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedListQueue
  def initialize
    @head = nil
    @tail = nil
  end

  # O(1)
  def enqueue(value)
    node = Node.new(value)

    if @head.nil?
      @head = node
      @tail = node

      return
    end

    @tail.next = node
    @tail = node
  end

  # O(1)
  def dequeue
    return unless @head

    @head.tap do
      @head = @head.next
      @tail = nil if @head.nil?
    end.value
  end
end

class LinkedListQueueTest < Test::Unit::TestCase
  def test_queue
    queue = LinkedListQueue.new

    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)

    assert_equal 1, queue.dequeue
    assert_equal 2, queue.dequeue
    assert_equal 3, queue.dequeue

    assert_nil queue.dequeue

    queue.enqueue(4)
    queue.enqueue(5)

    assert_equal 4, queue.dequeue
    assert_equal 5, queue.dequeue

    assert_nil queue.dequeue
  end
end

File: dsa/queues/golang/linked_list_queue.go
package main

import "fmt"

type Node struct {
  Value int
  Next *Node
}

type LinkedListQueue struct {
  Head *Node
  Tail *Node
}

func (queue *LinkedListQueue) Enqueue(element int) {
  node := &Node{Value: element}

  if queue.Head == nil {
    queue.Head = node
    queue.Tail = node
  } else {
    queue.Tail.Next = node
    queue.Tail = node
  }
}

func (queue *LinkedListQueue) Dequeue() *int {
  if queue.Head == nil {
    return nil
  }

  node := queue.Head
  queue.Head = node.Next

  if queue.Head == nil {
    queue.Tail = nil
  }

  return &node.Value
}

func main() {
  var element *int

  queue := &LinkedListQueue{}

  queue.Enqueue(1)
  queue.Enqueue(2)
  queue.Enqueue(3)

  for {
    if element = queue.Dequeue(); element != nil {
      fmt.Println(*element)
    } else {
      break
    }
  }

  queue.Enqueue(4)
  fmt.Println(*queue.Dequeue())
}
